_blogposts/compiler/overview-of-new_encoding.mdx

```diff
@@ -0,0 +1,245 @@
+Thus, the following code:
+
+```reasonml
+let u = Cons(1,Nil)
+```
+will generate the following JS output:
+
+```js
+var u = {_0: 1, _1 : /*Nil*/ 0 }; // No TAG data.
+```
+
+### Specialized for immutable list
+
+The `list` type is a built-in type; its type definition is similar to this :
+
+```reasonml
+type t ('a) = 
+    | []
+    | (::) ('a , t ('a))
+```
+
+Without any customization, this would generate JS objects with indexes like `_0`, `_1`, etc. 
+Since lists are so pervasive, we provide some special treatment so that
+
+```reasonml
+let u = [0,1,2,3]
+```
+Will generate js code as below:
+```js
+var u = {hd : 0, {tl : {hd : 1, {tl : {hd: 2, {tl : {hd :3 , tl : /*[]*/0 }}}}}}}
+```
+
+This is a minor change; we changed the name of `_0` to `hd` and `_1` to `tl`.
+
+### IC friendliness
+
+Types that have only one variant with a payload are in a perfect position for IC.
+
+The number of variants which does not carry payload will not affect IC, since the pattern match will do a split first. 
+
+Types with variants that have the same number of payloads, such as the red-black-tree example above, are also in a perfect position for IC.
+
+For other cases, it will hit a polymorphic IC in the V8 jit compiler, this is not the fastest running case.
+
+Note that you can always tweak the variant layout to make it IC friendly. For example, you can always introduce one level
+of indirection to make all variants share the same number of payloads:
+
+```reasonml
+type t = 
+    | A0 (a0) // 1 payload
+    | A1 (a1) // 1 payload
+    | A2 (a2) // 1 payload
+    | C0
+    | C1
+    | C2 // This will not affect IC 
+```
+
+### Variant in debug mode
+
+Note we only generate constructor names in comments for debugging.  
+When constructor names are attached to the data, it will be more useful for debugging. When debug mode is activated using `-bs-g`, 
+
+The generated code will be changed from below 
+```js
+var v0  = {TAG : 0/*Black*/, _0 : /*Empty*/ 0 , _1 : 3 , _2 : /*Empty */ 0};
+```
+to
+
+```js
+var v0  = {TAG : 0, _0 : /*Empty*/ 0 , _1 : 3 , _2 : /*Empty */ 0, [Symbol.for("name")]: "Black"};
+```
+
+## Polymorphic-variant (internal)
+
+Polymorphic variants allow users to use the types without declaring them first:
+
+```reasonml
+let u = 3 -> `hello
+```
+
+It will generate
+```js
+var u = {HASH : MAGIC_NUMBER, VAL: 3 }
+```
+The field of `HASH` is the hash of name `"hello"`, while the `VAL` is the payload
+
+### IC friendliness
+
+Polymorphic variants are always in a perfect position for IC, the compiler can ensure all generated objects are of the same shape. 
+This is because the payload is not unpacked; it is always just one payload.
+
+
+### Polymorphic variant in debug mode
+
+In debug mode, similar to variant, we carry the name in generated code for debugging,
+
+So instead of 
+```js
+var u = {HASH : MAGIC_NUMBER, VAL: 3 }
+```
+
+It will generate
+
+```js
+var u = {HASH : MAGIC_NUMBER, VAL: 3 , [Symbol.for("name")] : "hello"}
+```
```

Comment:
Outro

## Conclusion

BuckleScript users will soon profit from better runtime performance and a better debugging experience for various data types, such as variants, exceptions, lazy values and more. We discussed how we'll optimize our compile output for optimal inline caching behavior and showed off the new internal presentation of variant types as well.

We hope you like our newest changes, which will be released in our next major release 8.0.
Please stay tuned!